game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/RemoveUnprotectedUnits.java

```diff
@@ -0,0 +1,173 @@
+package games.strategy.triplea.delegate.battle.steps.change;
+
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.REMOVE_UNESCORTED_TRANSPORTS;
+
+import games.strategy.engine.data.Change;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.changefactory.ChangeFactory;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.Predicate;
+import lombok.AllArgsConstructor;
+import org.triplea.java.collections.CollectionUtils;
+
+@AllArgsConstructor
+public class RemoveUnprotectedUnits implements BattleStep {
+
+  private static final long serialVersionUID = 4357860848979564096L;
+
+  protected final BattleState battleState;
+
+  protected final BattleActions battleActions;
+
+  @Override
+  public List<String> getNames() {
+    final List<String> steps = new ArrayList<>();
+    if (battleState.getBattleSite().isWater()
+        && Properties.getTransportCasualtiesRestricted(battleState.getGameData())
+        && (battleState.getAttackingUnits().stream().anyMatch(Matches.unitIsTransport())
+            || battleState.getDefendingUnits().stream().anyMatch(Matches.unitIsTransport()))) {
+      steps.add(REMOVE_UNESCORTED_TRANSPORTS);
+    }
+    return steps;
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.REMOVE_UNPROTECTED_UNITS;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    checkAndRemoveUnits(bridge, BattleState.Side.DEFENSE);
+    checkAndRemoveUnits(bridge, BattleState.Side.OFFENSE);
+  }
+
+  public void checkAndRemoveUnits(final IDelegateBridge bridge, final BattleState.Side side) {
```

Comment:
This being a potential no-op if kinda concerning. I'm also wondering exactly what is meant by 'check', makes for a difficult public API.  Will this method ever be invoked if `(Properties.getTransportCasualtiesRestricted(battleState.getGameData()))` is false? Can we perhaps make it so that this step is never added to begin with? In that case you could assert that state to be true here

Reply:
Yes, it will be invoked if `getTransportCasualtiesRestricted` is false.  This method `checkAndRemoveUnits` is called from another step (see https://github.com/triplea-game/triplea/pull/7606/files#diff-009c223dfe7c716d07dc90f27899c933R1651).

The "check" is talking about the inner checks, not the `getTransportCasualtiesRestricted`.  It is checking for which units need to be removed.  Sometimes, units should be removed but most of the time, none are.

And the way that steps are now being created, they will always be added.  So we have to do all of the checks within the `execute` method or methods that it calls.

Reply:
Fair enough, could we rename this perhaps to something more direct like "removeUnprotectedUnits" and leave the fact that perhaps no units will be removed as an implementation detail. The API already does not expose any status flags that something was removed and we are purely modifying internal state of the class. Come to think about it, this is actually a code smell, some other class knows when this should be called, knows that this class needs to mutate its state rather than that being managed by this class itself. I think that is a bigger and more important issue, but perhaps a bit much to tackle now, I'd be happy if we can document the side-effects that happen and perhaps improve the name of the method.

Reply:
The removal of unprotected units and undefended transports is done in two different places.  I didn't want to duplicate the code so I have the code in this step and the other step calls this code as well.  I could create a helper class that does all this work as well and then have both steps call that helper class, if you'd like.

I'll rename it, though.